### ISSUE TITLE
Specifying alternative subgroups of parameters

### ISSUE THREAD

**Issue by lebrice** [Epistimio/orion#763]
Hey, I've got a weird question. 

I'm using the `HyperParameters` dataclass utility from `simple-parsing` to create my hyper-parameters.

I've created dataclasses for each group of hyper-parameters, for instance, one for each type of learning rate scheduler. (`StepLR` and `CosineAnnealingLR` below):
```python

from abc import abstractmethod
from typing import Any
from dataclasses import dataclass

from simple_parsing.helpers.hparams.hyperparameters import HyperParameters
from simple_parsing.helpers.hparams import uniform, log_uniform

from torch.optim.optimizer import Optimizer
from torch.optim.lr_scheduler import CosineAnnealingLR, StepLR, _LRScheduler


@dataclass
class LRSchedulerConfig(HyperParameters):
    """ Dataclass of hyper-parameters for creating a learning rate scheduler. """

    @abstractmethod
    def make_scheduler(self, optimizer: Optimizer) -> _LRScheduler:
        pass


@dataclass
class StepLRConfig(LRSchedulerConfig):
    step_size: int = uniform(10, 50, default=30)
    gamma: float = log_uniform(0.01, 0.5, default=0.1)

    def make_scheduler(self, optimizer: Optimizer) -> StepLR:
        return StepLR(optimizer, step_size=self.step_size, gamma=self.gamma)


@dataclass
class CosineAnnealingLRConfig(LRSchedulerConfig):
    T_max: int = uniform(80, 90, default=85, discrete=True)
    eta_min: float = log_uniform(1e-6, 1e-3, default=1e-5)

    def make_scheduler(self, optimizer: Optimizer) -> CosineAnnealingLR:
        return CosineAnnealingLR(optimizer, T_max=self.T_max, eta_min=self.eta_min)
```

These dataclasses are made to be easy to use with Orion:
```python
>>> StepLRConfig.get_orion_space_dict()
{'step_size': 'uniform(10, 50, discrete=True, default_value=30)', 'gamma': 'loguniform(1e-02, 0.5, default_value=0.1)'}
>>> StepLRConfig.sample()
StepLRConfig(step_size=29, gamma=0.25172643936909245)
```

Now, here is an example of dataclass that would contain all the hyper-parameters of a model.
```python
@dataclass
class MyModelHyperParameters(HyperParameters):
    # Learning rate
    learning_rate: float = log_uniform(1e-8, 1e-2, default=1e-3)
    # Batch size
    batch_size: int = log_uniform(16, 512, default=128, discrete=True)
    # Learning rate scheduler
    lr_scheduler: Union[CosineAnnealingLRConfig, StepLRConfig] = CosineAnnealingLRConfig()
```

Notice that it contains a field, `lr_scheduler`, which can **either** be a `CosineAnnealingLRConfig`, or a `StepLRConfig`.

**Now, here comes my question:**
```python
orion_space = MyModelHyperParameters.get_orion_space_dict()

# Should this give back:

# Option 1: Don't include the `lr_scheduler` in the space:
{
    "batch_size": "loguniform(16, 512, discrete=True, default_value=128)",
    "learning_rate": "loguniform(1e-08, 1e-02, default_value=0.001)",
}

# Option 2: Choose only one of the field types?
{
    "learning_rate": "loguniform(1e-08, 1e-02, default_value=0.001)",
    "batch_size": "loguniform(16, 512, discrete=True, default_value=128)",
    "lr_scheduler": {
        "T_max": "uniform(80, 90, discrete=True, default_value=85)",
        "eta_min": "loguniform(1e-06, 1e-03, default_value=1e-05)",
    },
}

# Option 3: Choose from some kind of union of two spaces?
{
        "learning_rate": "loguniform(1e-08, 1e-02, default_value=0.001)",
        "batch_size": "loguniform(16, 512, discrete=True, default_value=128)",
        "lr_scheduler": space_union([
              {
                      "step_size": "uniform(10, 50, discrete=True, default_value=30)",
                      "gamma": "loguniform(1e-02, 0.5, default_value=0.1)"
              },
              {
                       "T_max": "uniform(80, 90, discrete=True, default_value=85)",
                       "eta_min": "loguniform(1e-06, 1e-03, default_value=1e-05)",
              },
         ]),
}
```

By the way, as far as `sample()` goes, I've made it choose either one class or the other, and then sample from those priors.
```python
>>> MyModelHyperParameters.sample()
MyModelHyperParameters(learning_rate=0.002139543073257207, batch_size=155, lr_scheduler=StepLRConfig(step_size=27, gamma=0.28831543959169303))
>>> MyModelHyperParameters.sample()
MyModelHyperParameters(learning_rate=1.9865389375399682e-07, batch_size=31, lr_scheduler=CosineAnnealingLRConfig(T_max=85, eta_min=5.727805881168694e-05))
```
